### ISSUE TITLE
How to preprocess mask for multiclass segmentation?

### ISSUE THREAD

**Issue by JevgeniyVnk** [qubvel/segmentation_models#137]
Hello.
Images and masks in PNG format. 
8 Classes.
How to preprocess masks for categorical_crossentropy loss right?

I tried to use (128, 128, 1) mask, setting each pixel of the mask with a number from 0 to 7, but got error: 
Error when checking target: expected softmax to have shape (128, 128, 8) but got array with shape (128, 128, 1)

Code for model creation:
`BACKBONE = 'resnet50'
preprocess_input = get_preprocessing(BACKBONE)

model = Unet(BACKBONE, input_shape=(128,128,3),
             encoder_weights='imagenet',
             encoder_freeze=True,
            classes=8, activation='softmax')

model.compile('Adam', loss='categorical_crossentropy', metrics=[iou_score])`

Thanks. Sorry for bad English.

**Comment by JevgeniyVnk** [comment]
And how to convert NN prediction (128,128,8) to mask (128,128)? Should I apply argmax or something else?

**Comment by garelli** [comment]
Hello,

Thank you for sharing this project.

I have the same question in the case of multi-class segmentation.

**Comment by safak17** [comment]
Hello. :) Same problem here! 

`Error when checking target: expected sigmoid to have shape (224, 224, 4) but got array with shape (224, 224, 3).`

```python
model = Unet('vgg19', classes=4, input_shape=(224,224,3), encoder_weights='imagenet')
model.compile('Adam', loss="categorical_crossentropy", metrics=[iou_score])

from keras.preprocessing.image import ImageDataGenerator
datagen = ImageDataGenerator(
    featurewise_center=True,
    featurewise_std_normalization=True,
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    horizontal_flip=True)

datagen.fit(x_train)

model.fit_generator(
    datagen.flow(x_train, y_train, batch_size=32),
    steps_per_epoch=len(x_train) / 32, 
    epochs=1)
```

```python
# Here is the output of model.summary() function.
input_3 (InputLayer)            (None, 224, 224, 3)  0  
...
...
decoder_stage4_relu2 (Activatio (None, 224, 224, 16) 0           decoder_stage4_bn2[0][0]         
__________________________________________________________________________________________________
final_conv (Conv2D)             (None, 224, 224, 4)  580         decoder_stage4_relu2[0][0]       
__________________________________________________________________________________________________
sigmoid (Activation)            (None, 224, 224, 4)  0           final_conv[0][0]    
```

**Comment by garelli** [comment]
I have the answer; 

You need indeed to convert your RGB mask to a one-hot encoding image with shape (H,W,Channels) with Channels equals to the number of classes (containing the background).

Imagine you have an image/array (a mask) of shape (128,128,3). First you need to notice the unique elements which are corresponding to a label.

```
img.shape = (128,128,3)
list= []
for i in range(128):
    for j in range(128):
        list.append(img[i,j,:].tolist())

def uniq(lst):
    last = object()
    for item in lst:
        if item == last:
            continue
        yield item
        last = item

def sort_and_deduplicate(l):
    return list(uniq(sorted(l, reverse=True)))

palette = sort_and_deduplicate(list)
palette
```

```
[[128.0, 128.0, 0.0],
 [128.0, 0.0, 0.0],
 [0.0, 128.0, 0.0],
 [0.0, 0.0, 128.0],
 [0.0, 0.0, 0.0]]
```

palette is your current label pixels values in your RGB MASK (the unique elements)

Then, you can use a function like this to convert it to one-hot encoding:

```
one_hot_map = []
for colour in palette:
    class_map = tf.reduce_all(tf.equal(img2, colour), axis=-1)
    one_hot_map.append(class_map)
one_hot_map = tf.stack(one_hot_map, axis=-1)
one_hot_map = tf.cast(one_hot_map, tf.float32)

with tf.Session() as sess:
    print (sess.run(one_hot_map))

[[[0. 0. 0. 0. 1.]
  [0. 0. 0. 0. 1.]
  [0. 0. 0. 0. 1.]
  ...
  [0. 0. 0. 0. 1.]
  [0. 0. 0. 0. 1.]
  [0. 0. 0. 0. 1.]]
```

Here is for the processing of the masks (in this example you will have a (128,128,5) one-hot encoding mask.

And you can run 

# define model with output of N classes, where N > 1
model = Unet('resnet34', classes=5, activation='softmax')

# for multiclass segmentation choose another loss and metric
model.compile('Adam', loss='categorical_crossentropy', metrics=['categorical_accuracy'])


Once your model is trained, the predict function will outputs a (128,128,5) mask with probability inside it.

Then you can use argmax function with numpy to convert it to (128,128,3) true image and visualize results.

Hope it will helps

**Comment by JordanMakesMaps** [comment]
Haven't completed my project, but doing the following has helped get the ball rolling:

Each mask should be an 8-bit map (single channel), from [0:N] where 0 represents the background (even if there isn't one), and N is the number of classes. 

Send each mask into ```keras.utils.to_categorical()``` to get a one-hot-encoded representation of the label. Making your label the shape ```(batch_size, 128, 128, 5)```

**Comment by safak17** [comment]
| name       | dtype   | contains                                                                                                                               | shape                    | format | extension |
|------------|---------|----------------------------------------------------------------------------------------------------------------------------------------|--------------------------|--------|-----------|
| image      | uint8   | 0 to 255                                                                                                                               | 224x224x3                | RGB    | .jpg      |
| images     | uint8   | 2881 training images                                                                                                                   | 2881x224x224x3           |        |           |
| images_res | float32 | rescaled training images [0,1] interval                                                                                                | 2881x224x224x3 |        |           |
|            |         |                                                                                                                                        |                          |        |           |
| label      | uint8   | 0 to **NUM_CLASSES**<br> each number represents a class                                                                                    | 224x224                  | Gray   | .png      |
| labels     | uint8   | 2881 training labels                                                                                                                   | 2881x224x224             |        |           |
| labels_cat | float32 | Better to explain with [figure](https://user-images.githubusercontent.com/16006127/62009047-2f013e00-b163-11e9-9653-d61a6188cbec.png). | 2881x224x224x<br>**NUM_CLASSES** |        |           |

I thought, the whole _(not just for mask preprocessing)_ image&label preprocessing code for semantic segmentation task will be much cleaner for everyone _(especially for newbies)_.

```python
# Rescale the pixel values (between 0 and 255) to the [0, 1] interval. 
# "Neural networks prefer to deal with small input values." 
# Deep Learning with Python by Francois Chollet page 135

images_res = images.astype('float32')
images_res /= 255

labels_cat = to_categorical(y = labels,
                            num_classes = 4,
                            dtype = 'float32')
```

I have a very [imbalanced dataset](https://user-images.githubusercontent.com/16006127/62009613-f3b63d80-b169-11e9-9d5f-50a4afe3bbe3.png). And I tried to use **class_weight** and **sample_weight** as [Keras documentation](https://keras.io/models/sequential/) suggested but I didn't figure out how to use both two of them like [others](https://github.com/keras-team/keras/issues/3653). Finally, I applied the below code to tell the loss function that it should treat each class **equally**. Please, visit each link I provided above to understand better what I mean. I hope this will be useful for everyone.


```python
# For a single pixel value, I have NUM_CLASSES (in my case, it's 3) dimension.
# [0., 0., 1.] represents CLASS_C
# [0., 0., CLASS_WEIGHT] still represents CLASS_C 
# and also tells to loss function that, I care CLASS_C proportional to its CLASS_WEIGHT

CLASS_A_CHANNEL = 0
CLASS_B_CHANNEL = 1
CLASS_C_CHANNEL = 2

CLASS_A_WEIGHT = 1.02145046
CLASS_B_WEIGHT = 1000
CLASS_C_WEIGHT = 50

for label in labels_cat:
  label[ label[:,:,CLASS_A_CHANNEL] == 1 ] *= CLASS_A_WEIGHT
  label[ label[:,:,CLASS_B_CHANNEL] == 1 ] *= CLASS_B_WEIGHT
  label[ label[:,:,CLASS_C_CHANNEL] == 1 ] *= CLASS_C_WEIGHT
```

**Comment by JordanMakesMaps** [comment]
@safak17, here you have 4 classes in `to_categorical()` but in the second code block you mention you have three classes. I'm assuming the last class belongs to the background is ignored by training? How did you specify to Keras to ignore the background/void class within the masks so it's not involved in calculating the loss or made during predictions?

**Comment by safak17** [comment]
@JordanMakesMaps, I've made a typo and fixed it now. Sorry for the confusion. I didn't ignore the background/void class within the masks, I just try to **balance** the contribution of each class (including background) to the loss function.

| code block | int | one hot   | frequency  | weight         | weighted one hot       |
|------------|-----|-----------|------------|----------------|------------------------|
| CLASS_A    | 0   | [1, 0, 0] | 0.97886586 | 1.0215904353   | [1.0215904353, 0, 0]   |
| CLASS_B    | 1   | [1, 0, 0] | 0.00129868 | 770.0126282071 | [0, 770.0126282071, 0] |
| CLASS_C    | 2   | [0, 0, 1] | 0.01983546 | 50.414762249   | [0, 0, 50.414762249]   |

Here is the Python code that shows how I calculated the class frequencies. I just count the number of pixels belongs to each class, then divide it to the total number of pixels. For example, in the training set, **141501967 pixels belong to CLASS_A(0) class.**
```python

# y_train contains 224x224 label images (each one of them have INTEGER numbers 0,1,2)
# unique_elements = [0,                   1,          2]
# counts_elements = [141501967,      187733,    2867356]
# class_frequency = [0.97886586, 0.00129868, 0.01983546]
unique_elements, counts_elements = np.unique(ar = y_train,
                                             return_counts = True)

class_frequency = counts_elements / np.sum(counts_elements)
```
```python
# Step 1 : freq_class_a * CLASS_A_WEIGHT = 1
# Step 2 : 0.97886586 * CLASS_A_WEIGHT = 1
# Step 3 : CLASS_A_WEIGHT = 1.0215904353
```
To understand better `np.unique()` function please look at below image.
<img src="https://www.w3resource.com/w3r_images/python-numpy-image-exercise-94.png" alt="Drawing" width="400" height="250"/>

**Comment by JordanMakesMaps** [comment]
@safak17 thanks for the clarification. I realize now what I refer to as background is in fact a label I want to ignore completely. I believe that (in Keras) is denoted by a -1 or 255. Thanks!

**Comment by qubvel** [comment]
There is an example in `examples/` dir now

**Comment by aisamsoft** [comment]
> And how to convert NN prediction (128,128,8) to mask (128,128)? Should I apply argmax or something else?

you can simply use np.argmax(img, -1)

**Comment by shivarajkarki** [comment]
> 

Thank you.. Great solution
